File: src/main/java/com/adamkali/simpleide/browser/components/FileButton.java
package com.adamkali.simpleide.browser.components;

import com.adamkali.simpleide.Global;
import com.adamkali.simpleide.editor.EditorCoordinates;
import com.adamkali.simpleide.editor.io.OpenFile;
import com.adamkali.simpleide.preferences.EditorColors;
import com.adamkali.simpleide.project.SourceFile;

import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.util.function.Consumer;

public class FileButton extends JComponent {
    private final String name;
    private final int level;
    private SourceFile sourceFile;
    private Consumer<SourceFile> onFileClicked;
    private boolean hovered;

    public FileButton(String name) {
        this(name, 0);
    }

    public FileButton(SourceFile file, int level) {
        this(file.getFileName(), level);
        this.sourceFile = file;
        addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                if (onFileClicked != null && sourceFile != null) {
                    onFileClicked.accept(sourceFile);
                }
            }

            @Override
            public void mouseEntered(MouseEvent e) {
                hovered = true;
                repaint();
            }

            @Override
            public void mouseExited(MouseEvent e) {
                hovered = false;
                repaint();
            }
        });
    }

    public FileButton(String name, int level) {
        super();
        this.name = name;
        this.level = level;
        setFont(Global.getFont());
        setOpaque(true);
        setBackground(EditorColors.sidebarBackground());
        setForeground(EditorColors.sidebarForeground());
        setAlignmentX(LEFT_ALIGNMENT);
    }

    public void setOnFileClicked(Consumer<SourceFile> onFileClicked) {
        this.onFileClicked = onFileClicked;
    }

    public SourceFile getSourceFile() {
        return sourceFile;
    }

    public String getFileName() {
        return name;
    }

    public int getLevel() {
        return level;
    }

    private boolean isActiveFile() {
        return sourceFile != null
                && OpenFile.INSTANCE.getPath() != null
                && OpenFile.INSTANCE.getPath().equals(sourceFile.getPath());
    }

    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        if (isActiveFile()) {
            g.setColor(EditorColors.sidebarSelection());
        } else if (hovered) {
            g.setColor(EditorColors.sidebarHover());
        } else {
            g.setColor(EditorColors.sidebarBackground());
        }
        g.fillRect(0, 0, getWidth(), getHeight());

        FontMetrics fm = g.getFontMetrics();
        int textY = (getHeight() + fm.getAscent() - fm.getDescent()) / 2;
        int indent = EditorCoordinates.treeIndent(level);
        g.setColor(EditorColors.sidebarForeground());
        g.setFont(getFont());
        g.drawString(name, 16 + indent, textY);
    }

    @Override
    public Dimension getPreferredSize() {
        int indent = EditorCoordinates.treeIndent(level);
        int textWidth = Global.getStringWidth(name);
        int width = 16 + indent + textWidth + 8;
        return new Dimension(width, Global.getLineHeight() + 6);
    }

    @Override
    public Dimension getMinimumSize() {
        return new Dimension(80, Global.getLineHeight() + 6);
    }

    @Override
    public Dimension getMaximumSize() {
        return new Dimension(Integer.MAX_VALUE, getPreferredSize().height);
    }

    public static enum FileButtonState {
        IDLE,
        SELECTED
    }
}

File: src/main/java/com/adamkali/simpleide/browser/components/FileButton.kt
package com.adamkali.simpleide.browser.components

import com.adamkali.simpleide.Global
import com.adamkali.simpleide.editor.io.OpenFile
import com.adamkali.simpleide.preferences.EditorColors
import com.adamkali.simpleide.project.SourceFile
import java.awt.Component
import java.awt.Dimension
import java.awt.Graphics
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import javax.swing.JComponent

class FileButton(private val name: String, val level: Int = 0) : JComponent() {
    var sourceFile: SourceFile? = null
        private set
    private var onFileClicked: ((SourceFile) -> Unit)? = null
    private var hovered = false

    constructor(file: SourceFile, level: Int) : this(file.getFileName(), level) {
        sourceFile = file
    }

    val fileName: String
        get() = name

    init {
        font = Global.getFont()
        isOpaque = true
        background = EditorColors.sidebarBackground()
        foreground = EditorColors.sidebarForeground()
        alignmentX = Component.LEFT_ALIGNMENT
        addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(e: MouseEvent) {
                val file = sourceFile
                if (onFileClicked != null && file != null) {
                    onFileClicked?.invoke(file)
                }
            }

            override fun mouseEntered(e: MouseEvent) {
                hovered = true
                repaint()
            }

            override fun mouseExited(e: MouseEvent) {
                hovered = false
                repaint()
            }
        })
    }

    fun setOnFileClicked(onFileClicked: ((SourceFile) -> Unit)?) {
        this.onFileClicked = onFileClicked
    }

    private fun isActiveFile(): Boolean {
        val file = sourceFile ?: return false
        val openPath = OpenFile.path ?: return false
        return openPath == file.getPath()
    }

    override fun paintComponent(g: Graphics) {
        super.paintComponent(g)
        val selected = isActiveFile()
        val g2 = ExplorerItemStyle.prepare(g)
        ExplorerItemStyle.paintBackground(g2, width, height, hovered, selected)
        ExplorerItemStyle.paintFileIcon(g2, level, height, name)
        ExplorerItemStyle.paintLabel(g2, name, level, height, hovered || selected)
    }

    override fun getPreferredSize(): Dimension {
        return Dimension(
            ExplorerItemStyle.preferredWidth(level, name),
            ExplorerItemStyle.rowHeight()
        )
    }

    override fun getMinimumSize(): Dimension {
        return Dimension(80, ExplorerItemStyle.rowHeight())
    }

    override fun getMaximumSize(): Dimension {
        return Dimension(Integer.MAX_VALUE, preferredSize.height)
    }
}
